File: core/needle.py
"""Turning what an analyst TYPED into the bytes to search for.

One search box, several ways of spelling a needle. An analyst hunting a secret
does not always hold it as hex: it may be a literal string (``-----BEGIN``), a
wide string the process stored as UTF-16LE, a base64 blob pasted out of a
key-log or a JSON config, or a pointer value they read off another pane. Every
one of those is the same question — "where does this byte sequence live?" — and
:meth:`DumpSource.find_all` has always taken plain ``bytes``, so the only thing
that was ever hex-specific is this parse step.

WHY ONE MODULE. The hex normalisation used to be copy-pasted: ``tools_inspect``
had it, ``tools_pipeline._needle_from_key_hex`` had a VERBATIM copy whose
docstring said so, and ``region_analysis.parse_hex_pattern`` had a third. They
agreed only because a test
(``test_api_locate_key.test_hex_normalisation_matches_the_byte_search_box``)
held them together by hand. Adding five more formats to three copies is how
"the same string" starts meaning different bytes in different panes.

WHY ``auto`` RESOLVES TO HEX. ``dead``, ``cafe``, ``face`` and ``added`` are
all valid hex AND valid words. There is no reading of the input that is right
every time, so ``auto`` keeps the behaviour the box has always had — pure-hex
input is hex — and the ambiguity is surfaced instead of guessed at:
:func:`plausible_alternatives` names the other readings so a caller can offer a
one-click switch, and the web UI shows the resolved bytes before the search
runs. A wrong guess you can SEE costs one click; a wrong guess you cannot see
costs an analyst their afternoon.

WHY ``base64`` AND THE INTEGERS ARE NEVER AUTO-DETECTED. Their character sets
overlap plain text completely — ``Set-Cookie`` is as good a base64 charset run
as a real blob, and ``1234`` is as good a number as a string. Auto-detecting
them would make ``auto`` unpredictable, which is the one thing it must not be.
They are explicit-only, and reachable in one click through
:func:`plausible_alternatives`.
"""

import base64
import binascii
import re
from typing import List, Tuple

from .service_errors import CapabilityError, ErrorCategory

#: Every spelling the search box understands.
#:
#: Integer width and endianness live IN the format name rather than in extra
#: parameters, so the needle stays one ``(text, format)`` pair on all four
#: surfaces instead of growing width/endian fields on each of them.
NEEDLE_FORMATS: Tuple[str, ...] = (
    "auto",
    "hex",
    "text",
    "utf16le",
    "base64",
    "u32le",
    "u32be",
    "u64le",
    "u64be",
)

#: The concrete formats — everything :func:`detect_needle_format` may return.
CONCRETE_FORMATS: Tuple[str, ...] = tuple(
    f for f in NEEDLE_FORMATS if f != "auto"
)

#: ``{format: (byte width, byteorder)}`` for the fixed-width integer formats.
_INT_FORMATS = {
    "u32le": (4, "little"),
    "u32be": (4, "big"),
    "u64le": (8, "little"),
    "u64be": (8, "big"),
}

_HEX_RE = re.compile(r"\A[0-9a-fA-F]+\Z")
_BASE64_RE = re.compile(r"\A[A-Za-z0-9+/]+={0,2}\Z")

#: Shortest input worth offering as a base64 alternative.
#:
#: Short words are far too easy to read as base64 by accident (``dead`` decodes
#: to three bytes), so the hint would fire constantly and mean nothing.
_BASE64_HINT_MIN_CHARS = 16


def _invalid(message: str) -> CapabilityError:
    return CapabilityError(message, category=ErrorCategory.INVALID_INPUT)


def normalize_hex(text: str) -> str:
    """Strip whitespace and an optional leading ``0x``/``0X``.

    The spelling the byte-search box has always accepted: ``deadbeef``,
    ``0xdeadbeef``, ``de ad be ef`` and ``0XDEADBEEF`` all normalise to the
    same digits, because an analyst pastes the same string out of the hex
    viewer, a key-log line and a paper.
    """
    return "".join(text.split()).removeprefix("0x").removeprefix("0X")


def looks_like_hex(text: str) -> bool:
    """Is *text* an unambiguous, whole-byte hex string?

    Requires an EVEN digit count: an odd run is half a byte, and silently
    dropping or padding a nibble would search for something the analyst never
    typed.
    """
    cleaned = normalize_hex(text)
    return (
        len(cleaned) >= 2
        and len(cleaned) % 2 == 0
        and bool(_HEX_RE.match(cleaned))
    )


def looks_like_base64(text: str) -> bool:
    """Could *text* be a base64 blob worth OFFERING as an alternative?

    Deliberately conservative — see the module doc: this only ever feeds a
    hint, never a decision.
    """
    cleaned = "".join(text.split())
    return (
        len(cleaned) >= _BASE64_HINT_MIN_CHARS
        and len(cleaned) % 4 == 0
        and bool(_BASE64_RE.match(cleaned))
    )


def detect_needle_format(text: str) -> str:
    """Resolve ``auto`` to a concrete format.

    Hex when the input is unambiguously whole-byte hex, text otherwise. Never
    returns ``base64`` or an integer format — see the module doc.
    """
    if not text.strip():
        raise _invalid("Empty byte pattern")
    return "hex" if looks_like_hex(text) else "text"


def plausible_alternatives(text: str) -> List[str]:
    """Other formats *text* could sensibly be read as, best first.

    Powers the "also valid as …" hint beside the search box. Excludes whatever
    :func:`detect_needle_format` already chose, so the hint only ever offers a
    reading the user is NOT currently getting.
    """
    if not text.strip():
        return []
    resolved = detect_needle_format(text)
    out: List[str] = []
    if resolved != "hex" and looks_like_hex(text):
        out.append("hex")
    if resolved != "text":
        out.append("text")
    if looks_like_base64(text):
        out.append("base64")
    return out


def parse_needle(text: str, fmt: str = "hex") -> bytes:
    """Turn typed *text* into the bytes to search for.

    :param fmt: One of :data:`NEEDLE_FORMATS`; ``"auto"`` is resolved with
        :func:`detect_needle_format`.
    :raises CapabilityError: on an unknown format, an unparseable value, or an
        input that resolves to zero bytes.
    """
    if fmt not in NEEDLE_FORMATS:
        raise _invalid(
            f"Unknown pattern format: {fmt!r} "
            f"(expected one of {', '.join(NEEDLE_FORMATS)})")
    if not text.strip():
        raise _invalid("Empty byte pattern")
    if fmt == "auto":
        fmt = detect_needle_format(text)

    needle = _decode(text, fmt)
    if not needle:
        # An empty needle is not a query. `find_all_offsets` already returns []
        # for one (it must -- `mmap.find(b"", ...)` clamps and loops forever),
        # so without this the user would get a confident "0 hits" for an input
        # that was never searchable in the first place.
        raise _invalid(f"Empty byte pattern after decoding as {fmt}")
    return needle


def _decode(text: str, fmt: str) -> bytes:
    """Format-specific decode, assuming *fmt* is concrete and *text* non-blank."""
    if fmt == "hex":
        cleaned = normalize_hex(text)
        if not cleaned:
            raise _invalid("Empty byte pattern")
        try:
            return bytes.fromhex(cleaned)
        except ValueError as exc:
            # Name the likeliest cause rather than echoing binascii's wording:
            # an odd digit count is by far the most common way to get here, and
            # "non-hexadecimal number found" does not tell anyone that.
            hint = (
                " (odd number of hex digits — a byte needs two)"
                if len(cleaned) % 2 else ""
            )
            raise _invalid(
                f"Invalid hex byte pattern: {text!r}{hint}") from exc

    if fmt == "text":
        return text.encode("utf-8")

    if fmt == "utf16le":
        # No BOM: a needle is a fragment to find INSIDE a buffer, and a BOM
        # would only ever match at the start of a string the process stored
        # with one -- i.e. it would make the common case fail.
        return text.encode("utf-16-le")

    if fmt == "base64":
        cleaned = "".join(text.split())
        try:
            return base64.b64decode(cleaned, validate=True)
        except (binascii.Error, ValueError) as exc:
            raise _invalid(f"Invalid base64 pattern: {text!r}") from exc

    width, byteorder = _INT_FORMATS[fmt]
    stripped = text.strip()
    try:
        # base=0 so 0x/0o/0b prefixes work and a bare string stays decimal --
        # the same spelling `--offset` already accepts elsewhere in the CLI.
        value = int(stripped, 0)
    except ValueError as exc:
        raise _invalid(f"Invalid integer pattern: {text!r}") from exc
    if value < 0:
        raise _invalid(f"Integer pattern must not be negative: {text!r}")
    try:
        return value.to_bytes(width, byteorder)  # type: ignore[arg-type]
    except OverflowError as exc:
        raise _invalid(
            f"Integer pattern {stripped} does not fit in {width} bytes "
            f"({fmt})") from exc

File: frontend/src/utils/needle.ts
/**
 * Turning what an analyst TYPED into the bytes to search for.
 *
 * The TS half of a contract written twice. `core/needle.py` does the real
 * parse — it is what actually runs against the dump — but the search box has
 * to show the resolved bytes AS YOU TYPE, and a round trip per keystroke is
 * not an option. So the rules live here as well.
 *
 * `tests/fixtures/needle_vectors.json` is what stops the two drifting: the
 * same table is replayed by `tests/test_needle.py` and by
 * `tests/frontend/utils/needle.test.ts`. A divergence fails a test instead of
 * shipping a preview that disagrees with the search it just ran — which would
 * be worse than no preview at all, because the whole point of the preview is
 * that the user can trust it.
 *
 * WHY `auto` RESOLVES TO HEX. `dead`, `cafe` and `face` are valid hex AND
 * valid words. No reading is right every time, so `auto` keeps what the box
 * has always done — pure hex input is hex — and the ambiguity is SURFACED by
 * `plausibleAlternatives` rather than guessed at. A wrong guess you can see
 * costs one click.
 *
 * WHY base64 AND THE INTEGERS ARE NEVER AUTO-DETECTED. Their character sets
 * overlap ordinary text completely, so auto-detecting them would make `auto`
 * unpredictable — the one thing it must not be. They stay one click away.
 */

import { isHex, normalizeHex } from "@/utils/hex";

export const NEEDLE_FORMATS = [
  "auto",
  "hex",
  "text",
  "utf16le",
  "base64",
  "u32le",
  "u32be",
  "u64le",
  "u64be",
] as const;

export type NeedleFormat = (typeof NEEDLE_FORMATS)[number];

/** Everything `detectFormat` may return — every format except `auto`. */
export type ConcreteNeedleFormat = Exclude<NeedleFormat, "auto">;

/** `{format: [byte width, littleEndian]}` for the fixed-width integers. */
const INT_FORMATS: Record<string, [number, boolean]> = {
  u32le: [4, true],
  u32be: [4, false],
  u64le: [8, true],
  u64be: [8, false],
};

const BASE64_RE = /^[A-Za-z0-9+/]+={0,2}$/;

/**
 * Shortest input worth offering as base64.
 *
 * Short words are far too easy to read as base64 by accident (`dead` decodes
 * to three bytes), so a lower bound would make the hint fire constantly and
 * mean nothing.
 */
const BASE64_HINT_MIN_CHARS = 16;

/** Thrown for anything the user can fix by typing something else. */
export class NeedleError extends Error {}

/** Is this an unambiguous, WHOLE-byte hex string? */
export function looksLikeHex(text: string): boolean {
  const cleaned = normalizeHex(text);
  return cleaned.length >= 2 && cleaned.length % 2 === 0 && isHex(cleaned);
}

/** Could this be a base64 blob worth OFFERING? Never a decision, only a hint. */
export function looksLikeBase64(text: string): boolean {
  const cleaned = text.replace(/\s+/g, "");
  return (
    cleaned.length >= BASE64_HINT_MIN_CHARS &&
    cleaned.length % 4 === 0 &&
    BASE64_RE.test(cleaned)
  );
}

/** Resolve `auto` to a concrete format: hex when unambiguous, else text. */
export function detectFormat(text: string): ConcreteNeedleFormat {
  if (!text.trim()) throw new NeedleError("Empty byte pattern");
  return looksLikeHex(text) ? "hex" : "text";
}

/**
 * Other readings of `text`, best first — the "also valid as …" hint.
 *
 * Never includes the format already in effect: echoing the current reading
 * back is noise, and noise trains people to ignore the hint that matters.
 */
export function plausibleAlternatives(text: string): ConcreteNeedleFormat[] {
  if (!text.trim()) return [];
  const resolved = detectFormat(text);
  const out: ConcreteNeedleFormat[] = [];
  if (resolved !== "hex" && looksLikeHex(text)) out.push("hex");
  if (resolved !== "text") out.push("text");
  if (looksLikeBase64(text)) out.push("base64");
  return out;
}

/**
 * Parse typed text into the bytes that will be searched for.
 *
 * @throws NeedleError with a message meant for the user, not a console.
 */
export function parseNeedle(text: string, fmt: NeedleFormat = "hex"): Uint8Array {
  if (!text.trim()) throw new NeedleError("Empty byte pattern");
  const resolved = fmt === "auto" ? detectFormat(text) : fmt;
  const bytes = decode(text, resolved);
  if (bytes.length === 0) {
    // An empty needle is not a query: the scan returns no hits for one (it
    // must -- an empty needle makes an mmap search loop forever), so letting
    // it through would report a confident "0 hits" for something that was
    // never searchable.
    throw new NeedleError(`Empty byte pattern after decoding as ${resolved}`);
  }
  return bytes;
}

function decode(text: string, fmt: ConcreteNeedleFormat): Uint8Array {
  if (fmt === "hex") {
    const cleaned = normalizeHex(text);
    if (!cleaned) throw new NeedleError("Empty byte pattern");
    if (!isHex(cleaned)) {
      throw new NeedleError(`Invalid hex byte pattern: ${text}`);
    }
    if (cleaned.length % 2 !== 0) {
      throw new NeedleError(
        `Invalid hex byte pattern: ${text} (odd number of hex digits — a byte needs two)`,
      );
    }
    const out = new Uint8Array(cleaned.length / 2);
    for (let i = 0; i < out.length; i++) {
      out[i] = parseInt(cleaned.slice(i * 2, i * 2 + 2), 16);
    }
    return out;
  }

  if (fmt === "text") return new TextEncoder().encode(text);

  if (fmt === "utf16le") {
    // No BOM: a needle is a fragment to find INSIDE a buffer, so a BOM would
    // only ever match a string the process happened to store with one -- i.e.
    // it would make the common case fail.
    const out = new Uint8Array(text.length * 2);
    for (let i = 0; i < text.length; i++) {
      const code = text.charCodeAt(i);
      out[i * 2] = code & 0xff;
      out[i * 2 + 1] = code >> 8;
    }
    return out;
  }

  if (fmt === "base64") {
    const cleaned = text.replace(/\s+/g, "");
    if (!BASE64_RE.test(cleaned) || cleaned.length % 4 !== 0) {
      throw new NeedleError(`Invalid base64 pattern: ${text}`);
    }
    try {
      const binary = atob(cleaned);
      const out = new Uint8Array(binary.length);
      for (let i = 0; i < binary.length; i++) out[i] = binary.charCodeAt(i);
      return out;
    } catch {
      throw new NeedleError(`Invalid base64 pattern: ${text}`);
    }
  }

  const [width, littleEndian] = INT_FORMATS[fmt];
  const value = parseIntegerLiteral(text.trim());
  if (value < 0n) {
    throw new NeedleError(`Integer pattern must not be negative: ${text}`);
  }
  if (value >= 1n << BigInt(width * 8)) {
    throw new NeedleError(
      `Integer pattern ${text.trim()} does not fit in ${width} bytes (${fmt})`,
    );
  }
  const out = new Uint8Array(width);
  let rest = value;
  for (let i = 0; i < width; i++) {
    const byte = Number(rest & 0xffn);
    out[littleEndian ? i : width - 1 - i] = byte;
    rest >>= 8n;
  }
  return out;
}

/**
 * `BigInt` so a u64 keeps every bit.
 *
 * `Number` loses precision above 2^53, which is inside the range a u64 pointer
 * search routinely uses — the needle would be silently wrong in its low bytes.
 * Accepts the same spellings Python's `int(x, 0)` does, because an analyst
 * reads a pointer off one pane as hex and a length off another as decimal.
 */
function parseIntegerLiteral(text: string): bigint {
  const negative = text.startsWith("-");
  const body = negative ? text.slice(1) : text;
  if (!/^(0[xX][0-9a-fA-F]+|0[oO][0-7]+|0[bB][01]+|\d+)$/.test(body)) {
    throw new NeedleError(`Invalid integer pattern: ${text}`);
  }
  try {
    const value = BigInt(body);
    return negative ? -value : value;
  } catch {
    throw new NeedleError(`Invalid integer pattern: ${text}`);
  }
}

/** `de ad be ef` — the spelling the byte inspector and the legend already use. */
export function previewHex(bytes: Uint8Array): string {
  return Array.from(bytes, (b) => b.toString(16).padStart(2, "0")).join(" ");
}
